cmd/process-agent/collector.go

```diff
@@ -92,24 +97,83 @@ func (l *Collector) runCheck(c checks.Check, payloads chan checkPayload) {
 	messages, err := c.Run(l.cfg, atomic.AddInt32(&l.groupID, 1))
 	if err != nil {
 		log.Errorf("Unable to run check '%s': %s", c.Name(), err)
-	} else {
-		payloads <- checkPayload{messages, c.Name()}
-		// update proc and container count for info
-		updateProcContainerCount(messages)
-		if !c.RealTime() {
-			d := time.Since(s)
-			switch {
-			case runCounter < 5:
-				log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
-			case runCounter == 5:
-				log.Infof("Finished %s check #%d in %s. First 5 check runs finished, next runs will be logged every 20 runs.", c.Name(), runCounter, d)
-			case runCounter%20 == 0:
-				log.Infof("Finish %s check #%d in %s", c.Name(), runCounter, d)
-			}
+		return
+	}
+
+	for _, m := range messages {
+		body, err := encodePayload(m)
+		if err != nil {
+			log.Errorf("Unable to encode message: %s", err)
+			continue
+		}
+
+		extraHeaders := make(http.Header)
+		extraHeaders.Set(api.HostHeader, l.cfg.HostName)
+		extraHeaders.Set(api.ProcessVersionHeader, Version)
+		extraHeaders.Set(api.ContainerCountHeader, strconv.Itoa(getContainerCount(m)))
+
+		if cid, err := clustername.GetClusterID(); err == nil && cid != "" {
+			extraHeaders.Set(api.ClusterIDHeader, cid)
+		}
+
+		if err := l.acquireQueueBytes(len(body)); err != nil {
+			log.Errorf("Unable to enqueue message: %s", err)
+			continue
+		}
+
+		payload := checkPayload{
+			name:    c.Name(),
+			body:    body,
+			headers: extraHeaders,
+		}
+		select {
+		case payloads <- payload:
+		default:
+			bytesEnqueued := atomic.LoadInt32(&l.bytesEnqueued)
+			log.Warnf("Delivery queue is full.  Size=%d, Bytes Enqueued=%d.  Waiting for space...", len(payloads), bytesEnqueued)
+
+			payloads <- payload
```

Comment:
This is an interesting choice - I would've expected to prioritize data at the tail by dropping messages from the head of our queue if we reached the queue size/length limits.

What do the other agents do in this scenario?

Reply:
The main agent forwarder has (what appears to me) a more complicated retry loop but it does appear to prefer newer data, albeit only 30 messages (by default).

However the forwarder strictly drops messages by time assuming that one message is the same as any other. 

However, since we are imposing a memory constraint, the messages are no longer equivalent.  For example, would it be worth it to drop 3 older process messages in exchange for 1 newer network message?  If we want to go this route, its a worth a longer conversation about how we want to prioritize the data.

In any case, per the comment on the size of this queue, it should allow enqueuing ~100 minutes of payloads before sends would block here.